Feature subscriptions

yamahaxes opened the pull request.

yamahaxes committed:
feat: add stats-server, stats-dto modules

yamahaxes committed:
feat: add stats-client module

yamahaxes committed:
feat: change stats-client

yamahaxes committed:
feat: add dockerfile

yamahaxes committed:
feat: change pom

yamahaxes committed:
feat: add ewm-service and ewm-db to docker-compose.yml

yamahaxes committed:
feat: check style

yamahaxes committed:
refactor: remove class

yamahaxes committed:
feat: Добавить контроллеры CategoryController

yamahaxes committed:
feat: Добавить контроллеры UserController

yamahaxes committed:
feat: Добавить контроллер EventControllerAdm #1

yamahaxes committed:
feat: Добавить контроллер EventControllerAdm #2

yamahaxes committed:
feat: Добавить контроллер EventControllerPub

yamahaxes committed:
feat: Добавить контроллер EventControllerPvt #1

yamahaxes committed:
feat: Исправить для тестов Postman #1

yamahaxes committed:
feat: Реализовать эндпоинт Request

yamahaxes committed:
feat: Реализовать эндпоинты в приватном EventControllerPvt по работе с request

yamahaxes committed:
feat: Event postman tests

yamahaxes committed:
refactor: #1

yamahaxes committed:
feat: реализовать эндпоинт Compilation

yamahaxes committed:
refactor: Compilation postman test

yamahaxes committed:
refactor: all postman test success

yamahaxes committed:
refactor: all postman test success #2

yamahaxes committed:
refactor: codestyle

yamahaxes committed:
refactor: docker settings

yamahaxes committed:
refactor: #1

yamahaxes committed:
refactor: #2

yamahaxes committed:
refactor: #3

yamahaxes committed:
refactor: Codereview #1

yamahaxes committed:
refactor: Codereview #1. SendStat убрать отправление отдельным потоком. (не проходит тесты)

yamahaxes committed:
feat: Subscription endpoint

yamahaxes committed:
feat: get actual events by subscription

yamahaxes committed:
feat: add privacy User

yamahaxes committed:
refactor: remove Custom mappers

yamahaxes committed:
feat: Исправление ошибок

yamahaxes committed:
feat: Добавление тестов postman

yamahaxes committed:
feat: Добавление тестов postman

yamahaxes committed:
feat: Исправление тестов Postman

yamahaxes committed:
refactor: изменение README.md

sergeymamonov commented on stats-server/src/main/java/ru/practicum/ewm/stats/server/controller/StatsServerController.java:
```diff
@@ -0,0 +1,43 @@
+    @GetMapping("/stats")
+    public ResponseEntity<List<HitDtoResponse>> getStats(@RequestParam @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm:ss") LocalDateTime start,
+                                                         @RequestParam @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm:ss") LocalDateTime end,
+                                                         @RequestParam(required = false) Optional<List<String>> uris,
```
Основное название Optional это обертки над результатом. Не рекомендуется использовать Optional для обертки над входными параметрами. Подробнее можно почитать здесь https://struchkov.dev/blog/optional-in-java/#%D0%BC%D0%B5%D1%82%D0%BE%D0%B4-orelsethrowsupplier-s

sergeymamonov commented on ewm-service/src/main/java/ru/practicum/ewm/event/controller/pub/EventControllerPub.java:
```diff
@@ -0,0 +1,59 @@
+                                         @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm:ss") @RequestParam(required = false) LocalDateTime rangeStart,
+                                         @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm:ss") @RequestParam(required = false) LocalDateTime rangeEnd,
+                                         @RequestParam(required = false) Boolean onlyAvailable,
+                                         @RequestParam(required = false) Optional<EventSort> sort,
```
Аналогично, здесь лучше не использовать Optional

sergeymamonov commented on ewm-service/src/main/java/ru/practicum/ewm/event/service/EventService.java:
```diff
@@ -0,0 +1,338 @@
+
+        Sort sort = Sort.unsorted();
+        if (sortOptional.isPresent()) {
+            switch (sortOptional.get()) {
```
Лучше дописать ветку default.

sergeymamonov commented on ewm-service/src/main/java/ru/practicum/ewm/event/service/EventService.java:
```diff
@@ -0,0 +1,338 @@
+
+    private void sendStat(HttpServletRequest request) {
+
+        //new Thread(() -> {
```
В проекте не должно быть никакого закомментированного кода.

sergeymamonov commented on ewm-service/src/main/java/ru/practicum/ewm/util/EwmUtils.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class EwmUtils {
```
Не используйте рефклексию в промышленном коде. Рефлексия работает очень медленно, нарушает принципы ООП. Если вы выбираете рефлексию для промышленного кода, значит вы что-то делаете неверно.

sergeymamonov reviewed (changes requested):
Сейчас есть три модуля в проекте stats-client, stats-dto и stats-server. Лучше их переложить в модуль stats, как это указывалось в задании на первом этапе разработки ТЗ. У модуля stats родительским помником будет корневой помник проекта. А у stats-client, stats-dto и stats-server родительским помником будет помник stats.
Посмотрите про использование Optional. Его не рекомендуется использовать во входных параметрах. Подсветил пару мест в коде.

yamahaxes committed:
refactor: Исправить замечания ревью #1

yamahaxes committed:
refactor: Убрать рефлексию ревью #1

sergeymamonov commented on ewm-service/src/main/java/ru/practicum/ewm/event/service/EventService.java:
```diff
@@ -0,0 +1,334 @@
+                case VIEWS:
+                    sort = Sort.by("views"); break;
+                default:
+                    sort = Sort.unsorted();
```
Может лучше кинуть ошибку? Ведь получили не поддерживаемый вариант для сортировки.

sergeymamonov commented on ewm-service/src/main/java/ru/practicum/ewm/event/service/EventService.java:
```diff
@@ -0,0 +1,334 @@
+            } else {
+                log.info("STAT: error created hit={}, status={}", dto, result.getStatusCode());
+            }
+        } catch (RuntimeException ex) {
```
По-хорошему нужно в случае, если мы не получили HttpStatus.CREATED делать ретрай запроса. И ограничить количество ретраев каким-то количеством попыток.

yamahaxes committed:
refactor: Выкинуть исключение, когда SORT некорректный.
feat: Добавить попытки отправки статистики. Добавить обновление просмотров (views) события.

yamahaxes committed:
refactor: codestyle review
